src/main/java/org/jahia/modules/appshell/Main.java

```diff
@@ -119,6 +153,46 @@ public String getContextPath() {
         }
     }
 
+    private String getSiteKey(HttpServletRequest request) {
+
+        String siteKey = null;
+
+        // first let's try to resolve the site from a request parameter if it exists
+        if (request.getParameter("site") != null && request.getParameter("site").length() < 100) {
+            siteKey = request.getParameter("site");
+            // we validate the site parameter to protect against attacks.
+            if (ServicesRegistry.getInstance().getJahiaSitesService().getSitesNames().contains(siteKey)) {
+                return siteKey;
+            }
+        }
+
+        // now let's try to resolve it from the domain
+        if (!Url.isLocalhost(request.getServerName())) {
+            try {
+                siteKey = ServicesRegistry.getInstance().getJahiaSitesService().getSitenameByServerName(request.getServerName());
+                if (siteKey != null) {
+                    return siteKey;
+                }
+            } catch (JahiaException e) {
+                logger.error("Error resolving site key from domain", e);
+            }
+        }
+
+        // finally let's try to resolve it from URL patterns configured in the applications.
+        for (Pattern urlPattern : urlPatterns) {
+            Matcher urlMatcher = urlPattern.matcher(request.getPathInfo());
+            if (urlMatcher.matches()) {
+                siteKey = getGroupValue(urlMatcher, "siteKey");
+                if (siteKey == null || siteKey.trim().isEmpty()) {
+                    String workspace = getGroupValue(urlMatcher,"workspace");
+                    Locale locale = LocaleUtils.toLocale(getGroupValue(urlMatcher,"language"));
+                    siteKey = getSiteFromNode(getGroupValue(urlMatcher,"nodeUUID"), workspace, locale);
+                }
+            }
```

Comment:
I think it miss a return statement in case a pattern matched, we dont want to continue iterating on others patterns ?